python/minetester/minetest_env.py

```diff
@@ -0,0 +1,505 @@
+        headless: bool = True,
+        verbose_logging: bool = False,
+    ):
+        if config_dict is None:
+            config_dict = {}
+        self.unique_env_id = str(uuid.uuid4())
+
+        self.display_size = DisplaySize(*display_size)
+        self.fov_y = fov
+        self.fov_x = self.fov_y * self.display_size.width / self.display_size.height
+        self.render_mode = render_mode
+        self.headless = headless
+        self.start_xvfb = start_xvfb and self.headless
+        self.gameid = gameid
+        assert gameid is not None or world_dir is not None, (
+            "Either a gameid or a world_dir must be provided!",
+        )
+
+        assert (
+            not self.start_xvfb or sys.platform == "linux"
+        ), "Xvfb is only supported on Linux."
+
+        if render_mode == "human":
+            self._start_pygame()
+
+        # Define action and observation space
+        self._configure_spaces()
+
+        # Define Minetest paths
+        self._set_artifact_dirs(
+            artifact_dir, world_dir, config_path
+        )  # Stores minetest artifacts and outputs
+        self.minetest_executable = Path(minetest_executable)
+        assert (
+            self.minetest_executable.exists()
+        ), f"Minetest executable not found: {self.minetest_executable}"
+
+        # ZMQ port
+        self.env_port = get_free_port() if env_port is None else env_port
+        self.verbose_logging = verbose_logging
+
+        # ZMQ objects
+        self.socket = None
+        self.context = None
+
+        # Minetest processes
+        self.client_process = None
+
+        # Env objects
+        self.last_obs = None
+        self.render_fig = None
+        self.render_img = None
+
+        # Seed the environment
+        self.base_seed = base_seed
+        self.world_seed = world_seed
+        # If no world_seed is provided
+        # seed the world with a random seed
+        # generated by the RNG from base_seed
+        self.reseed_on_reset = world_seed is None
+        self.seed(self.base_seed)
+
+        # Write minetest.conf
+        self.config_dict = config_dict
+        self._write_config()
+
+        # Configure logging
+        logging.basicConfig(
+            filename=os.path.join(self.log_dir, f"env_{self.unique_env_id}.log"),
+            filemode="a",
+            format="%(asctime)s,%(msecs)d %(name)s %(levelname)s %(message)s",
+            datefmt="%H:%M:%S",
+            level=logging.DEBUG,
+        )
+
+        # Start X server virtual frame buffer
+        self.default_display = x_display or 0
+        if "DISPLAY" in os.environ:
+            self.default_display = int(os.environ["DISPLAY"].split(":")[1])
+        self.x_display = x_display or self.default_display
+        self.xserver_process = None
+        if self.start_xvfb:
+            self.x_display = x_display or self.default_display + 4
+            self.xserver_process = start_xserver(self.x_display, self.display_size)
```

Comment:
I think it's simpler to use `xvfb-run --auto-servernum minetest` rather than start the x server and then later start minetest. Any reason not to do it that way?